pkg/codegen/go/gen.go

```diff
@@ -39,9 +39,35 @@ import (
 )
 
 type typeDetails struct {
-	ptrElement   bool
-	arrayElement bool
-	mapElement   bool
+	input      bool
+	ptrInput   bool
+	arrayInput bool
+	mapInput   bool
+
+	output      bool
+	ptrOutput   bool
+	arrayOutput bool
+	mapOutput   bool
+}
+
+func (d *typeDetails) mark(input, output bool) {
+	d.input = d.input || input
+	d.output = d.output || input || output
```

Comment:
Why do you set `output` to true if `input` is true? Are you planning to remove unreferenced output types later on?

Reply:
Generating the Go form of an input type requires that we also generate the Go form of the corresponding output type. This is due to the way we define input type interfaces, which reference output types:

```
type StringInput interface {
 	pulumi.Input

 	ToStringOutput() StringOutput
 	ToStringOutputWithContext(context.Context) StringOutput
 }
```